File: frontend/src/lib/security/redaction.ts
const DEFAULT_SENSITIVE_KEYS = new Set([
  "authorization", "access_token", "refresh_token", "id_token", "token", "password",
  "secret", "cookie", "set-cookie", "api_key", "apikey", "session", "sessionid",
]);

function isSensitiveKey(key: string): boolean {
  const normalized = key.toLowerCase().replaceAll("-", "_");
  return DEFAULT_SENSITIVE_KEYS.has(normalized) || normalized.endsWith("_token") || normalized.endsWith("_secret");
}

export function redactSensitive<T>(input: T): T {
  return redactValue(input, new WeakSet<object>()) as T;
}

function redactValue(input: unknown, seen: WeakSet<object>): unknown {
  if (input === null || typeof input !== "object") return input;
  if (seen.has(input)) return "[Circular]";
  seen.add(input);
  if (Array.isArray(input)) return input.map((value) => redactValue(value, seen));
  const output: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(input)) {
    output[key] = isSensitiveKey(key) ? "[REDACTED]" : redactValue(value, seen);
  }
  return output;
}

export function scrubText(input: string): string {
  return input
    .replace(/Bearer\s+[A-Za-z0-9._~+\/-]+=*/gi, "Bearer [REDACTED]")
    .replace(/(access_token|refresh_token|id_token|api[_-]?key)=([^&\s]+)/gi, "$1=[REDACTED]");
}

File: backend/src/deskpilot_llmops/redaction.py
from __future__ import annotations

import re
from collections.abc import Mapping

_SECRET_KEYS = {"authorization", "access_token", "refresh_token", "id_token", "password", "secret", "cookie", "set-cookie"}
_BEARER = re.compile(r"(?i)bearer\s+[a-z0-9._~+\-/]+=*")
_EMAIL = re.compile(r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b", re.I)


def redact_value(value: object, *, key: str = "") -> object:
    if key.lower() in _SECRET_KEYS or any(token in key.lower() for token in ("token", "password", "secret")):
        return "[REDACTED]"
    if isinstance(value, Mapping):
        return {str(k): redact_value(v, key=str(k)) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [redact_value(v) for v in value]
    if isinstance(value, str):
        return _EMAIL.sub("[EMAIL_REDACTED]", _BEARER.sub("Bearer [REDACTED]", value))
    return value
